core/solver/ir.cpp

```diff
@@ -160,21 +160,31 @@ std::unique_ptr<LinOp> Ir<ValueType>::conj_transpose() const
 
 template <typename ValueType>
 void Ir<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    this->apply_with_initial_guess(b, x, this->get_default_initial_guess());
+}
+
+
+template <typename ValueType>
+void Ir<ValueType>::apply_with_initial_guess(const LinOp* b, LinOp* x,
+                                             initial_guess_mode guess) const
 {
     if (!this->get_system_matrix()) {
         return;
     }
+    this->prepare_initial_guess(b, x, guess);
```

Comment:
By pulling this into the Lambda, we would no longer need to call `run` on it, since we already know its value type. In that case, we could even eliminate the function entirely by pulling its functionality into `apply_dense_impl` as I suggested above.